preprocessors/ner/ner.py

```diff
@@ -0,0 +1,219 @@
+
+    # Run NER tagger on words
+    words =  ner_tagger.tag(words)
+    
+    if only_ner:
+        rtn  = []
+        for i in words:
+            if len(i[1]) > 4:
+                rtn.append(i)
+        return rtn
+    
+    return words
+
+
+"""
+Function to find the index of a given substring (without whitespace) in a string, interms of words starting with index 1.
+e.g. find_index("Hello I'm Namdar.", "Namdar") = 3
+:text: the string to check
+:word: the word to check for
+"""
+def find_index(text, word):
+    index = 1
+    for i in text.split():
+        if word in i:
+            return index
+        index += 1
+    return -1
+
+
+@app.route('/preprocessor', methods=['POST', 'GET'])
+def main():
+    
+    logging.debug("Received request")
+
+    with open('./schemas/preprocessors/ner.schema.json') as jsonfile:
+        data_schema = json.load(jsonfile)
+
+    with open('./schemas/preprocessor-response.schema.json') as jsonfile:
+        schema = json.load(jsonfile)
+
+    with open('./schemas/definitions.json') as jsonfile:
+        definition_schema = json.load(jsonfile)
+
+    with open('./schemas/request.schema.json') as jsonfile:
+        first_schema = json.load(jsonfile)
+
+    schema_store = {
+        data_schema['$id']: data_schema,
+        schema['$id']: schema,
+        definition_schema['$id']: definition_schema
+    }
+
+    resolver = jsonschema.RefResolver.from_schema(
+        schema, store=schema_store)
+
+    content = request.get_json()
+
+    try:
+        validator = jsonschema.Draft7Validator(first_schema, resolver=resolver)
+        validator.validate(content)
+    except jsonschema.exceptions.ValidationError as e:
+        logging.error(e)
+        return jsonify("Invalid Preprocessor JSON format"), 400
+
+    # ------ START COMPUTATION ------ #
+
+    name_ = "1"
+    captions = {}                            # dict to store the captions
+    text = get_alt(content)                  # extract alt text and save it in the caption dict
+    captions[name_] = text
+    html_ = content['graphic'] # get the image html content
+    
+    # save pic and caption to directory for clipscore evaluation
+    with open(CONTEXT_DIR+'/captions.json', "w") as outfile:
+        json.dump(captions, outfile)
+    save_image(html_, name_, IMAGE_DIR)
+
+    # check if we have an alt text
+    if len(text) < 2:
+        logging.info("No alttxt")
+        return "", 204
+
+    # create path parameters for clipscore
+    parameters = Namespace(candidates_json=CONTEXT_DIR+'/captions.json', compute_other_ref_metrics=1, image_dir=IMAGE_DIR, references_json=None, save_per_instance=CONTEXT_DIR+'/score.json')
+    
+    # calculate the clipscore
+    score = round(clipscore.main(parameters)['1']['CLIPScore'], 3)
+    # compute the NERs
+    ners = stanford_ner(text)
+    ner_data = []
+    for i in ners:
+        my_dict = {}
+        my_dict['value'] = i[0]
+        my_dict['tag'] = i[1]
+        my_dict['index'] = find_index(text, i[0])
```

Comment:
Worried about the efficiency of this `find_index` function being called for each item rather than iterating once...not blocking for now, just noting.

Reply:
I rewrote it in the new code, the new approach also solves the problem with had with repeated tokens.